src/Compilers/Core/Portable/Binding/BindingDiagnosticBag.cs

```diff
@@ -207,23 +207,29 @@ internal void AddDependencies(CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo)
             }
         }
 
+        internal bool Add(Location location, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo)
+            => Add(location, useSiteInfo, static location => location);
+
         internal bool Add(SyntaxNode node, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo)
-        {
-            return Add(node.Location, useSiteInfo);
-        }
+            => Add(node, useSiteInfo, static node => node.Location);
 
-        internal bool AddDiagnostics(SyntaxNode node, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo)
-        {
-            return AddDiagnostics(node.Location, useSiteInfo);
-        }
+        internal bool Add(SyntaxToken token, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo)
+            => Add(token, useSiteInfo, static token => token.GetLocation());
 
         internal bool AddDiagnostics(Location location, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo)
+            => AddDiagnostics(location, useSiteInfo, static location => location);
+
+        internal bool AddDiagnostics(SyntaxNode node, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo)
+            => AddDiagnostics(node, useSiteInfo, static node => node.Location);
+
+        internal bool AddDiagnostics<TData>(TData data, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo, Func<TData, Location> getLocation)
```

Comment:
>internal bool AddDiagnostics<TData>(TData data, CompoundUseSiteInfo<TAssemblySymbol> useSiteInfo, Func<TData, Location> getLocation) [](http://example.com/codeflow?start=8&length=132)

It might be worth having an overload of this helper that takes ```Location``` as the original method used to do. In order to avoid allocating delegates for those consumers.